docs/architecture/adr-012-ibc-relayer-v2.md

```diff
@@ -0,0 +1,744 @@
+    pub key_entry: KeyEntry,
+}
+
+pub struct FullCosmosChainContext {
+    pub handle: ChainHandle,
+    pub signer: Signer,
+    pub tx_config: TxConfig,
+    pub key_entry: KeyEntry,
+    pub batch_channel: CosmosBatchChannel,
+    pub telemetry: CosmosTelemetry,
+}
+```
+
+A visual inspection of the type definitions above shows that the full-featured
+Cosmos chain context contains two additional fields, `batch_channel` and
+`telemetry`. This is because the full preset makes use of the batched message
+sender and telemetry components, which requires the chain context to provide
+the batch channels and telemetry context. Hence, if a user want to use the
+full-featured Cosmos relayer, they would also have to instantiate and provide
+the additional parameters when constructing the chain context.
+
+The Cosmos chain context is implemented as an MVP for the relayer v1.5. As a
+result, it delegates most of the chain operations to the existing `ChainHandle`
+code in relayer v1. As we progress toward the v2 relayer, the Cosmos chain
+context is expected to eventually remove its dependency to `ChainHandle`, and
+directly holds the low-level fields such as `grpc_address`.
+
+To make use of the Cosmos chain context for relaying between Cosmos chains,
+we implement the one-for-all traits for our Cosmos chain contexts roughly as
+follows:
+
+```rust
+impl OfaBaseChain for MinCosmosChainContext
+{
+    type Preset = MinimalPreset;
+
+    type Error = CosmosError;
+
+    type Height = CosmosHeight;
+
+    type Timestamp = CosmosTimestamp;
+
+    type Message = CosmosMessage;
+
+    type Event = CosmosEvent;
+
+    type ChainStatus = CosmosChainStatus;
+
+    /* ... */
+
+    async fn query_chain_status(&self) -> Result<CosmosChainStatus, CosmosError> {
+        /* ... */
+    }
+
+    async fn send_messages(
+        &self,
+        messages: Vec<CosmosMessage>,
+    ) -> Result<Vec<Vec<CosmosEvent>>, CosmosError> {
+        /* ... */
+    }
+
+    /* ... */
+}
+```
+
+For demonstration purpose, the above code is slightly simplified from the actual
+Cosmos chain implementation. Readers are encouraged to refer to the
+`ibc-relayer-cosmos` itself to see the full implementation details.
+
+In the `OfaBaseChain` implementation for `MinCosmosChainContext`, we first
+define the `Preset` type attribute to `MinimalPreset`, indicating that we
+only want to build a minimal relayer with our concrete context. We then bind
+the abstract types such as `Error` and `Height` to the Cosmos-specific types
+such as `CosmosError` and `CosmosTimestamp`. Finally, we implement the abstract
+methods that are reqruired to be provided by the concrete chain implementation,
+such as `query_chain_status` and `send_messages`.
+
+We would also do the same implementation for the concrete relay contexts, such
+as implementing `OfaBaseRelay` for  `MinCosmosRelayContext`, which would contain
+two `MinCosmosChainContext`s. Once the traits are implemented, the relayer
```

Comment:
It's not clear why there would be two `MinCosmosChainContext`s.

Reply:
Made clarification that one is for the source chain, and the other is for the destination chain.